draft-irtf-cfrg-vdaf.md

```diff
@@ -1363,6 +1376,57 @@ class PrgSha3(Prg):
 ~~~
 {: title="Definition of PRG PrgSha3."}
 
+### PrgFixedKeyAes128 {#prg-fixed-key-aes128}
+
+While PrgSha3 as described above can be securely used in all cases where a Prg
+is needed in the VDAFs described in this document, there are some cases where
+a more efficient instantiation based on fixed-key AES is possible. For now, this
+is limited to the Prg used inside the Idpf {{idpf}} implementation in Poplar1
+{{idpf-poplar}}. It is NOT RECOMMENDED to use this Prg anywhere else.
+See Security Considerations {{security}} for a more detailed discussion.
+
+~~~
+class PrgFixedKeyAes128(Prg):
+    # Associated parameters
+    SEED_SIZE = 16
+
+    def __init__(self, seed, custom, binder):
+        self.length_consumed = 0
+
+        # Use SHA-3 to derive a key from the binder and customization
+        # strings. Note that the AES key does not need to be kept
+        # secret from any party. However, when used with IpdfPoplar,
+        # we require the binder to be a random nonce.
+        #
+        # Implementation note: This step can be cached across PRG
+        # evaluations with many different seeds.
+        self.fixed_key = cSHAKE128(binder, 16, b'', custom)
```

Comment:
Does this really need to use a hash function that barely anyone has?  I know that ... shiny... but this is a bit of a pain.

Reply:
Do you mean SHA-3 or cSHAKE in particular?

We're using SHA-3 because we needed an XOF. We looked at using HKDF-SHA-256 as an alternative, but the performance was pretty bad. Plus, we have some cases where we need longer outputs than HKDF permits.

We're using cSHAKE for PrgSha3 as well. What's nice about it is that we get distinct notions of "customization" and "binder" strings. (See https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-05#section-6.2.3.)

Reply:
FWIW, @simon-friedberger may be planning to implement this in NSS :)

Reply:
We could use SHA-2 here if for example you waned PrgFixedKeyAes128 but not PrgSha3.

Reply:
Simon already reached out about the nasty complexity that this produces for us.  I think that the current approach involves a bunch of non-NSS crypto, which is sort-of fine, but suboptimal.

I realize that there are advantages to using the new stuff, but that doesn't make it less awkward.

Reply:
Oh, I should say, why are you concerned about the performance of something that you instantiate so few times during a protocol run?  We're using HKDF with SHA-256 in our prototypes of IPA and it's in the noise, even for tiny runs.

Reply:
The reason I chose cSHAKE here was to be consistent with PrgSha3 and not require yet another hash function dependency. If this is creating problems for implementations, I don't have a problem with switching to HKDF-SHA-256, or even simply SHA-256/128. We just want the AES key to be random for each client and don't require anything else from the hash function.

Reply:
It's reasonable to rely on the same primitive here. If we ever allow more options for the PrgSha3 this should be updated as well but it's fine for now.

Reply:
@martinthomson the performance issue with HKDF comes up when we need an XOF, like we need in Prio3. (See https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/106#issuecomment-1397492095.) I agree the performance cost would be negligible, and I also think plain SHA-2 would be fine.